### ISSUE TITLE
mingw64-cmake build error

### ISSUE THREAD

**Issue by napasa** [tesseract-ocr/tesseract#1709]
command: mkdir build && cd build && mingw64-cmake .. && make
system:Linux localhost.localdomain 4.16.7-200.fc27.x86_64  Fedora release 27 (Twenty Seven)

distribution
CMakeFiles/tesseract.dir/objects.a(tesseractmain.cpp.obj):tesseractmain.cpp:(.text+0x13a2): undefined reference to `TIFFSetWarningHandler'
CMakeFiles/tesseract.dir/objects.a(tesseractmain.cpp.obj):tesseractmain.cpp:(.text+0x1478): undefined reference to `tesseract::Dict::GlobalDawgCache()'
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/tesseract.dir/build.make:88: bin/tesseract.exe] Error 1
make[1]: *** [CMakeFiles/Makefile2:68: CMakeFiles/tesseract.dir/all] Error 2
make: *** [Makefile:130: all] Error 2

**Comment by zdenop** [comment]
Seems like missing linking of tiff library (in case of `#if defined(HAVE_TIFFIO_H) ) && defined(_WIN32)`)...
Is it still issue with the current code?

**Comment by zdenop** [comment]
Can you please retest? We plan to release 4.00 version soon.

**Comment by stweil** [comment]
I suggest to use an automake build instead of cmake. That works.

**Comment by stweil** [comment]
@zdenop, no need to fix this for 4.0.0 as automake works.

**Comment by zdenop** [comment]
Automake tools on windows are pain... I have it on my old computer and I skipped it  on new and installed cmake. But not mingw64 ;-)
IMO it should work already or fix should be easy. But we need somebody how test it. If not I will close the issue.

**Comment by stweil** [comment]
The initial report was for a Fedora Linux host, so it was a cross build. That's how I build the Tesseract installer for Windows, too.

When building on a Windows host, I use Cygwin64 and run a cross build targeting Mingw-w64. It's much slower than on a Linux host, but otherwise that combination works fine.

**Comment by zdenop** [comment]
Closing as no reaction from reporter. If issues is still valid, please retest with the latest master

**Comment by aleksey-nikolaev** [comment]
Yep, I catch this bug.
Qt 5.12 + mingw73_64
The problem is - you are trying to call tesseract::Dict::GlobalDawgCache() which is not exported from dll. I'm realy messed, why even MSVC didn't catch this bug, maybe due to it is a static function and somehow linker could get the definition, also I don't remember that this bug was appearing a year ago even with MinGW.
So, just add `TESS_API` to the `Dict`: `src\dict\dict.h`

       class TESS_API Dict {

Probably other objects may need to be checked for missing TESTS_API.

**Comment by zdenop** [comment]
What it get that error? Please provide all details.

**Comment by aleksey-nikolaev** [comment]
@zdenop, the error is the same as that of @napasa, who created this issue, otherwise I would create a new one.

```
Scanning dependencies of target tesseract
[ 99%] Building CXX object CMakeFiles/tesseract.dir/src/api/tesseractmain.cpp.obj
[100%] Linking CXX executable bin\tesseract.exe
CMakeFiles\tesseract.dir/objects.a(tesseractmain.cpp.obj):tesseractmain.cpp:(.text.startup+0x253): undefined reference to `tesseract::Dict::GlobalDawgCache()'
collect2.exe: error: ld returned 1 exit status
mingw32-make.exe[2]: *** [CMakeFiles\tesseract.dir\build.make:87: bin/tesseract.exe] Error 1
```

**Comment by zdenop** [comment]
I am not interesting in error message: I need steps to reproduce it. 
You probably use unsupported compiler on windows - see wiki about [supported compiler](https://github.com/tesseract-ocr/tesseract/wiki/Compiling#window). + I am able to compile it with clang.

**Comment by aleksey-nikolaev** [comment]
Install Qt 5.12.3 with mingw 7.3; CMake; Pkg-Config. Configure environments. Build:
```
cmake ../tesseract-source -G"CodeBlocks - MinGW Makefiles"  -DCMAKE_INSTALL_PREFIX=%INSTALL_PATH%
cmake --build . --target install --config %BUILD_CONFIGURATION%
```
I am also able to compile it with MSVC, clang, gcc on unix, but not with [mingw](https://mingw-w64.org/doku.php).

I can't understand why MinGW might be an unsupported compiler of tesseract? The header of this issue contain "mingw64-cmake" and no one says that it is unsupported, the wiki also contains a few words about [mingw](https://github.com/tesseract-ocr/tesseract/wiki/Compiling#mingwmsys). Аnd [Qt used mingw for build](https://doc-snapshots.qt.io/qt5-5.12/supported-platforms.html#windows)

Here https://github.com/tesseract-ocr/tesseract/blob/master/src/dict/dict.h#L87
we have a class that is not [exported from DLL](http://docs.microsoft.com/cpp/cpp/dllexport-dllimport). And here https://github.com/tesseract-ocr/tesseract/blob/master/src/api/tesseractmain.cpp#L626 the call of function that not exported from DLL. And even you have not defined the macro `TESS_API` some compilers (clang, gcc on unix) will work fine.

**Comment by zdenop** [comment]
1. I we have list of supported compilers (E.g. there are people who maintained support for them). Making list of not supported compilers or platforms is insane - it would be never completed ;-)
2. TESS_API macro should effect IMO only "visibility build" (hiding of functions), so it should not effect standard build. So if clang, MSVC and gcc works as expected and MinGW not, I guess there is problem (in MinGW or cmake compiler instructions).
3. From my experience Dict::GlobalDawgCache() is not needed (as public function) I am not sure why it was used in tesseract executable. I need to check how it fill effects API.
4. Thanks for details about configuring project - I will try it later.

**Comment by zdenop** [comment]
I reopened issue because this topic (at least) produces some questions:

- I was able to reproduce error with mingw on windows, gcc on linux + MSVC, and clang with MSVC compatibility on windows have no problem with this setup. Does mingw use by default `-fvisibility=hidden -fvisibility-inlines-hidden`?
- I plan to test clang with gcc/mingw compatibility.  
- Usage of GlobalDawgCache() in tesseract main was introduced by @stweil in commit https://github.com/tesseract-ocr/tesseract/commit/78142593d24edcf9fcc7fc83e14d39733b55aa41. I am not sure if memory leak mentioned in commint is related to report like "`ObjectCache(0x7fe81e8fcee0)::~ObjectCache(): WARNING! LEAK! object 0x1abf240 still has count 1 (id /usr/local/share/....`". If yes, that calling [api->End()](https://github.com/tesseract-ocr/tesseract/wiki/APIExample#basic-example) should be (right/better?) solution.
- TODO: test `--enable-visibility` autoconf build

**Comment by aleksey-nikolaev** [comment]
One more call of unexported function github.com/tesseract-ocr/tesseract/blob/master/src/api/tesseractmain.cpp#L639
IMHO, it is a very bad idea to use something from the library that is not explicitly 
 marked for export, even if export is transparent for some platform.

**Comment by stweil** [comment]
[MinGW ](http://www.mingw.org/)is not well maintained, and there is [Mingw-w64](http://mingw-w64.org/doku.php) as an alternative. Tesseract builds fine with Mingw-w64. I use it regularly for cross builds from Cygwin and from Linux. Therefore I don't think that Tesseract must support MinGW.

**Comment by aleksey-nikolaev** [comment]
@stweil , the main question is why `tesseract.exe` called the functions that are not explicitly marked for export from `tesseract.dll`. For newer toolchain this is only a logical problem, and most of them didn't show even a warning, but some of them used opaque export https://docs.microsoft.com/cpp/cpp/dllexport-dllimport
And here we got a strange thing: MSVC feel fine for static function call drom dll.
I use Qt with MinGW 7.3 https://doc-snapshots.qt.io/qt5-5.12/supported-platforms.html#windows i.e. Mingw-w64.

**Comment by zdenop** [comment]
try now: should be solved with 539673b5036b2b4b9ece14af4a7614f9e65824c2

**Comment by aleksey-nikolaev** [comment]
it's ok for now.
